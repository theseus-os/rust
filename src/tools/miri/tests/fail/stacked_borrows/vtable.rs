//@error-pattern: vtable pointer does not have permission
#![feature(ptr_metadata)]

trait Foo {}

impl Foo for u32 {}

fn uwu(thin: *const (), meta: &'static ()) -> *const dyn Foo {
    core::ptr::from_raw_parts(thin, unsafe { core::mem::transmute(meta) })
}

fn main() {
    unsafe {
        let orig = 1_u32;
        let x = &orig as &dyn Foo;
        let (ptr, meta) = (x as *const dyn Foo).to_raw_parts();
        let _ = uwu(ptr, core::mem::transmute(meta));
    }
}
