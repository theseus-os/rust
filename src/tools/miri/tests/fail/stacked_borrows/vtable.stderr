error: Undefined Behavior: constructing invalid value: encountered vtable pointer does not have permission to read drop function pointer
  --> RUSTLIB/core/src/ptr/metadata.rs:LL:CC
   |
LL |     unsafe { PtrRepr { components: PtrComponents { data_address, metadata } }.const_ptr }
   |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ constructing invalid value: encountered vtable pointer does not have permission to read drop function pointer
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
   = note: backtrace:
   = note: inside `std::ptr::from_raw_parts::<dyn Foo>` at RUSTLIB/core/src/ptr/metadata.rs:LL:CC
note: inside `uwu` at $DIR/vtable.rs:LL:CC
  --> $DIR/vtable.rs:LL:CC
   |
LL |     core::ptr::from_raw_parts(thin, unsafe { core::mem::transmute(meta) })
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside `main` at $DIR/vtable.rs:LL:CC
  --> $DIR/vtable.rs:LL:CC
   |
LL |         let _ = uwu(ptr, core::mem::transmute(meta));
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

note: some details are omitted, run with `MIRIFLAGS=-Zmiri-backtrace=full` for a verbose backtrace

error: aborting due to previous error

